---
title: "s_figure_11_carbapenem_exposure"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```


```{r environment, echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse","gridExtra",'kableExtra',"tableone","ape","ggtree")

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")
source("./lib/common_functions.R") 
```

```{r}
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile") 
# GWAS table
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS")
gwas_mat <- readRDS("./data/GWAS/hits/gwas_mat.RDS") 
# Matrices
## Kleborate matrix
kleborate <- readRDS("./data/GWAS/matrices/kleborate_mat.RDS") %>% as.data.frame 
## KPC Plasmids 
KPC_containing_clusters <- readRDS("./data/GWAS/matrices/KPC_containing_cluster_mat.RDS") %>% as.data.frame
## Core genome burden models
models <- c("model1","model1.2","model2","model2.2","model3","model3.2")
for(x in models){
  grouped_model <- paste0(x,"_burden.RDS")
  assign(gsub(".RDS","",grouped_model),readRDS(paste0("./data/GWAS/matrices/",grouped_model)) %>% as.data.frame)
}
## BLBLI clustering
pclustering <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS") 
blbli_asr <- readRDS("./data/asr_clustering/blbli_asr.RDS")
## Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
## Depth
depth <- readRDS("./data/KPC_coverage/normalized/KPNIH1_KPC_plasmid_depth_by_feature_median_chromosome_normalized.RDS")
# Join
df <- left_join(df,kleborate %>% mutate(isolate_no = rownames(.)))
df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.)))  
df <- left_join(df,pclustering %>% select(isolate_no,blbli_asr_cluster_renamed))
df$blbli_asr_cluster_renamed_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible" )
   
df <- left_join_matrix(model1_burden,df,"_1")
df <- left_join_matrix(model1.2_burden,df,"_1.2")
df <- left_join_matrix(model2_burden,df,"_2")
df <- left_join_matrix(model2.2_burden,df,"_2.2")
df <- left_join_matrix(model3_burden,df,"_3")
df <- left_join_matrix(model3.2_burden,df,"_3.2")
df <- left_join(df,canonical %>% mutate(isolate_no = rownames(.))) 
df <- left_join(df,depth %>% mutate(isolate_no = rownames(.)))
```

```{r}
df$blbli_asr_cluster_simplified <- ifelse(df$blbli_asr_cluster_renamed=="No Feature","Susceptilbe",ifelse(df$blbli_asr_cluster_renamed=="Singleton","Singleton","Cluster"))
df$clade_I <- as.factor(df$clade_I )
df$clade_I <- relevel(df$clade_I, ref = "Clade II")
df_CI <- df %>% subset(clade_I == "Clade I")
df_CII <- df %>% subset(clade_I != "Clade I")
```

```{r}
known_hits <-  c("OmpK36_putative_function_altering","OmpK36_promoter_IS",'OmpK36_truncation_kleborate',"PBP_any","RamA")  

## Known Hits
df$known <- get_pattern(known_hits,df) %>% .$presence
novel_synchronous <-  gwas_table %>% subset(resistance_category != "Modulator" & nn_qc=="Yes") %>% subset(!locus_tag %in% c("KPNIH1_RS18665","KPNIH1_RS07805")) %>% .$genotype
df$novel_synchronous <- get_pattern(novel_synchronous,df) %>% .$presence
df$AA552 <- get_pattern("AA552",df) %>% .$presence
df$known_ST258_GWAS <- get_pattern(c(known_hits,novel_synchronous),df) %>% .$presence
df$known_ST258_GWAS_AA552 <- get_pattern(c(known_hits,novel_synchronous,"AA552"),df) %>% .$presence
df$blbli_dich_num <- as.numeric(df$blbli_dich_num)
```


# Figure 1 - by clade and clustering
## By NS and clade
```{r,width=12.5}

alluvium_plot_data <- df %>% select(blbli_dich,carbapenem_1d,third_plus_ceph_1d,tigecycline_1d,known,clade_I,AA552,blbli_asr_cluster_simplified)
alluvium_plot_data$carbapenem <- ifelse(alluvium_plot_data$carbapenem_1d == 1,"History","No History")
alluvium_plot_data$ceph <- ifelse(alluvium_plot_data$third_plus_ceph_1d == 1,"History","No History")
alluvium_plot_data$tige <- ifelse(alluvium_plot_data$tigecycline_1d == 1,"History","No History")
alluvium_plot_data$genotype <- ifelse(alluvium_plot_data$AA552 == 1 ,"AA552 KPC plasmid",ifelse(alluvium_plot_data$known == 1,"Carbapenem resistance-associated genotype","None"))

FA <- ggplot(data=alluvium_plot_data,aes(x=carbapenem,fill=as.factor(clade_I))) + facet_grid(~blbli_dich,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Carbapenem Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5))+ ylim(NA,275) + clade_colors_scale

FE <- ggplot(data=alluvium_plot_data,aes(x=ceph,fill=as.factor(clade_I))) + facet_grid(~blbli_dich,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Cephalosporin Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5)) +clade_colors_scale + ylim(NA,275)
 
``` 
## By phyloaware and clade
```{r}
FB <- ggplot(data=alluvium_plot_data,aes(x=carbapenem,fill=as.factor(clade_I))) + facet_grid(~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Carbapenem Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5))+ ylim(NA,250) + clade_colors_scale

FF <- ggplot(data=alluvium_plot_data,aes(x=ceph,fill=as.factor(clade_I))) + facet_grid(~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Cephalosporin Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5)) +clade_colors_scale+ ylim(NA,275) 
 

```
## By phyloaware and clade and genotype
```{r}
FC <- ggplot(data=alluvium_plot_data,aes(x=carbapenem,fill=as.factor(genotype))) + facet_grid(~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Carbapenem Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5))+ ylim(NA,250) + genotype_scale

FG <- ggplot(data=alluvium_plot_data,aes(x=ceph,fill=as.factor(genotype))) + facet_grid(~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Cephalosporin Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5)) +genotype_scale + ylim(NA,250)
```
## By phyloaware and clade and genotype
```{r}
FD <- ggplot(data=alluvium_plot_data,aes(x=carbapenem,fill=as.factor(genotype))) + facet_grid(clade_I~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Carbapenem Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5))+ ylim(NA,200) + genotype_scale

FH <- ggplot(data=alluvium_plot_data,aes(x=ceph,fill=as.factor(genotype))) + facet_grid(clade_I~blbli_asr_cluster_simplified,scales = 'free',drop = F) +  geom_bar()+ theme_bw_me + xlab("Prior Cephalosporin Exposure") + ylab("Frequency") +
  theme(panel.margin=unit(.05, "lines"),
        panel.border = element_rect(color = "black", fill = NA, size = 0.5), 
        strip.background = element_rect(color = "black", size = 0.5)) +genotype_scale + ylim(NA,225)
```

# Plot

```{r,fig.width=15,fig.height=7.5}
wo_legend <- cowplot::plot_grid(FA + theme(legend.position="none"),FB+ theme(legend.position="none"),FC+ theme(legend.position="none"),FD+ theme(legend.position="none"),FE+ theme(legend.position="none"),FF+ theme(legend.position="none"),FG+ theme(legend.position="none"),FH+ theme(legend.position="none"),ncol=4,labels =c("A.","B.","C.","D.","E.","F.","G.","H."), label_size = 18,rel_widths = c(1,1.25,1.25,1.4,1,1.25,1.25,1.4))
legend <- cowplot::plot_grid(get_legend(FA),get_legend(FC),ncol=2)
s_figure_11 <- cowplot::plot_grid(wo_legend,legend,ncol=1,rel_heights = c(1,.075))
s_figure_11
```

```{r,message=F,echo=F,error=F,warning=F,fig.width=27.5,fig.height=15} 
ggsave(plot = s_figure_11, filename = './figures/s_figure_11.png', width = 15, height = 7.5,limitsize=FALSE)
```