---
title: "Supplemental Figure 3"
output: html_document
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
---

# Combination Therapy Manuscript
```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r environment, echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse","cowplot","phyloAMR","ggtree","gontjesr","ggnewscale",'gridExtra')

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./lib/common_functions.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")
```

# Data
```{r}
df <- readRDS("./data/dataset/df.RDS") 
# Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
df <- left_join(df,canonical)

# tn4401 data
tn4401 <- readRDS("./data/TETyper/TETyper_curated.RDS")  
df <- left_join(df,tn4401)

# load the kleborate
kleborate <- readRDS("./data/kleborate/kleborate_curated.RDS") %>% as.data.frame
df <- left_join(df,kleborate)
```

```{r}
tr <- read.tree("./data/tree/tree.treefile")  
nn <- readRDS("./data/nearest_neighbor_comparisons/nn_comparisons.RDS") 
blbli_asr <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS")
df <- df %>% left_join(.,blbli_asr)
df$blbli_clustering_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible")
```

```{r}
drug_entry <- c('OmpK35-25%','OmpK36_c25t','OmpK36GD','OmpK36TD','OmpK36_truncation_kleborate','OmpK36_non_syn','OmpK36_putative_function_altering','OmpK36_intergenic','OmpK36_promoter_IS')
drug_exit <- c("AcrAB_TolC_any",'RamR',"RamA")
target_modification <- c('PBP2',"PBP4")

carbapenem <- c(drug_entry,drug_exit,target_modification)
```


# FA - phylogeny 
```{r ,fig.height=15,fig.width=15,echo=F,message=F,error=F,warning=F} 
rownames(df) <- df$isolate_no 
 
consistent_theme_sfig3 <- theme(legend.position = 'bottom',legend.direction="horizontal", legend.justification = "center", legend.key = element_rect(colour = c('black')),legend.box.spacing = unit(.00001, "cm"),legend.key.size = unit(0.25, "cm"),legend.key.width = unit(0.25, "cm") , legend.title = element_text(size=12),legend.text = element_text(size=10),legend.title.align=0.5,legend.text.align = 0, legend.margin=margin(t=-1,r=0,b=0,l=0,unit="cm"),legend.spacing.x=unit(.25, "cm")) 

  #Step #1: Clades
  p.1 <- gheatmap(ggtree(tr),df %>% select(clade_I) %>% `colnames<-`("ST258 Clade"), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.05)   + clade_colors_scale_v + consistent_theme_sfig3

  #Step #3: BL/BLI Cluster
  p.2 <- p.1 + new_scale_fill()
  p.3 <-  gheatmap(p.2,df %>% select(blbli_asr_cluster_renamed) %>% mutate_all(as.factor) %>% `colnames<-`("BL/BLI Clustering"), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.05,offset =.0000125) + cluster_scale + consistent_theme_sfig3

  #Step #4: MVB Binary
  p.4 <- p.3 + new_scale_fill()
  p.5 <-  gheatmap(p.4,df %>% select(blbli_dich) %>% `colnames<-`(c("BL/BLI Resistance")), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.05,offset =.000025) + resistance_scale + consistent_theme_sfig3

  #Step #6: MVB MIC
  p.6 <- p.5 + new_scale_fill()
  p.7 <-  gheatmap(p.6,df %>% select(MVB_log_2,IR_log_2) %>%   mutate_all(as.factor)  %>%`colnames<-`(c("MVB MIC","IR MIC")), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.1,offset =.0000375) + Log2_scale + consistent_theme_sfig3

  # Step #7: Add tn4401 allele
  p.7.1 <- p.7 + new_scale_fill() 
  p8 <-  gheatmap(p.7.1,df %>% select(tn4401_allele) %>% mutate_all(as.factor)  %>% `colnames<-`("tn4401 isoform"), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.05,offset =.0000625)  + tn4401_scale + consistent_theme_sfig3 + ylim(NA,480)

# KPC allele
  p8.1 <- p8 + new_scale_fill()
  p9 <-  gheatmap(p8.1,df %>% select(KPC_type) %>% mutate_all(as.factor)  %>% `colnames<-`("KPC allele"), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.05,offset =.000075)  +kpc_scale + consistent_theme_sfig3 + ylim(NA,480)
0
# relevant alleles
  p9.1 <- p9 + new_scale_fill()
  s_figure_3_A <-  gheatmap(p9.1,df %>% select(any_of(carbapenem)) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.,),'OmpK35-25'='ompK35 porin truncation at 25%','OmpK36_c25t'='25 cytosine-to-thymine transition in ompK36','OmpK36_L3_mutations'='Loop 3 insertion in ompK36','OmpK36GD'='GD loop 3 insertion','OmpK36TD'='TD loop 3 insertion','OmpK36_truncation_kleborate'='ompK36 truncation', 'OmpK36_non_syn'='ompK36 non-synonymous mutation','OmpK36_putative_function_altering'='ompK36 PFAV','OmpK36_intergenic'='Mutation in intergenic region of ompK36','OmpK36_promoter_IS'='Insertion seqeunce at ompK36 promoter',"AcrAB_TolC_any" = 'acrAB-tolC efflux pump mutant','RamR'='ramR efflux pump regulator PFAV',"RamA"='ramA efflux pump activator PFAV','PBP_any'='Non-synonymous mutations in penicillin-binding-proteins','PBP2'='Penicillin-binding protein-2 mutant',"PBP4"='Penicillin-binding protein-4 mutant')), colnames_position = "top",colnames_angle=90, colnames_offset_y = 0.25, hjust = 0, color = NA, font.size = 5.5, width = 0.7,offset =.0000875) + feature_scale_v  + consistent_theme_sfig3 + ylim(NA,645)
  s_figure_3_A
```

# FB - non freq s
```{r,fig.height=7.5,fig.width=10}
variables_to_use <- carbapenem %>% subset(.!='OmpK35-25%')

gwas_nons_freq <- variants_freq(variables_to_use,df)   
gwas_nons_freq$locus_tag<- factor(gwas_nons_freq$locus_tag, levels=rev(unique(variables_to_use)))
                                  
gwas_nons_freq$`Significant Hit` <- recode(gwas_nons_freq$locus_tag ,'OmpK35-25'='ompK35 porin truncation at 25%','OmpK36_c25t'='25 cytosine-to-thymine transition in ompK36','OmpK36_L3_mutations'='Loop 3 insertion in ompK36','OmpK36GD'='GD loop 3 insertion','OmpK36TD'='TD loop 3 insertion','OmpK36_truncation_kleborate'='ompK36 truncation', 'OmpK36_non_syn'='ompK36 non-synonymous mutation','OmpK36_putative_function_altering'='ompK36 PFAV','OmpK36_intergenic'='Mutation in intergenic region of ompK36','OmpK36_promoter_IS'='Insertion seqeunce at ompK36 promoter',"AcrAB_TolC_any" = 'acrAB-tolC efflux pump mutant','RamR'='ramR efflux pump regulator PFAV',"RamA"='ramA efflux pump activator PFAV','PBP_any'='Non-synonymous mutations in penicillin-binding-proteins','PBP2'='Penicillin-binding protein-2 mutant',"PBP4"='Penicillin-binding protein-4 mutant')  

s_figure_3_descriptive_plot_theme <-  theme(legend.position="bottom",axis.text = element_text(size=12,colour = "black"),axis.title = element_text(size=14,colour = "black"),legend.title = element_text(size=16 ,colour = "black"),legend.text = element_text(size=14,colour = "black"),legend.title.align=0.5 ,legend.key.size = unit(0.5, "cm"),legend.key.width = unit(0.5, "cm")) 

s_figure_3_B <- freq_plot(gwas_nons_freq) +  s_figure_3_descriptive_plot_theme
s_figure_3_B
```

# FC - nn data 
```{r,fig.height=7.5,fig.width=10}
# C. FC MIC (MVB * IR overlay) 
nn_data_melt <- lapply(variables_to_use,generate_nn_melt_data,nn_data=nn) %>% do.call(rbind,.)
nn_data_melt$locus_tag <- factor(nn_data_melt$locus_tag, levels=rev(unique(variables_to_use)))  
nn_data_melt$locus_tag<- factor(nn_data_melt$locus_tag, levels=rev(unique(variables_to_use)))
                                  
nn_data_melt$locus_tag<- recode(nn_data_melt$locus_tag ,'OmpK35-25'='ompK35 porin truncation at 25%','OmpK36_c25t'='25 cytosine-to-thymine transition in ompK36','OmpK36_L3_mutations'='Loop 3 insertion in ompK36','OmpK36GD'='GD loop 3 insertion','OmpK36TD'='TD loop 3 insertion','OmpK36_truncation_kleborate'='ompK36 truncation', 'OmpK36_non_syn'='ompK36 non-synonymous mutation','OmpK36_putative_function_altering'='ompK36 PFAV','OmpK36_intergenic'='Mutation in intergenic region of ompK36','OmpK36_promoter_IS'='Insertion seqeunce at ompK36 promoter',"AcrAB_TolC_any" = 'acrAB-tolC efflux pump mutant','RamR'='ramR efflux pump regulator PFAV',"RamA"='ramA efflux pump activator PFAV','PBP_any'='Non-synonymous mutations in penicillin-binding-proteins','PBP2'='Penicillin-binding protein-2 mutant',"PBP4"='Penicillin-binding protein-4 mutant') 
     
s_figure_3_C <- nn_plot(nn_data_melt)   + s_figure_3_descriptive_plot_theme
s_figure_3_C
```

```{r,fig.height=12.5,fig.width=19.75}
s_figure_3BC <- plot_grid(s_figure_3_B,s_figure_3_C ,nrow = 2,labels = c("B","C"),label_size = 22,align = "hv",label_x = -0.02)

s_figure_3 <- plot_grid(s_figure_3_A,s_figure_3BC,labels = c("A",""),label_size = 22,ncol = 2,rel_heights = c(1,1),rel_widths = c(1,0.625)) 

s_figure_3
```

```{r,message=F,echo=F,error=F,warning=F,fig.height=12.5,fig.width=19.75} 
ggsave(plot = s_figure_3, filename = './figures/s_figure_3.png', height = 12.5, width = 19.75,  limitsize=FALSE, bg='white')
```
 