---
title: "Supplemental figure 4: All significant genome-wide association study hits for β-lactam/β-lactamase inhibitor resistance"
output: html_document
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
---

# Combination Therapy Manuscript
```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

# Environment
```{r}
#Load Packages
packages <- c("tidyverse","cowplot","ggtree","ggnewscale","ape","phytools","phyloAMR","gridExtra")
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./lib/common_functions.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")

# Print
Sys.info()
sessionInfo() 

# Set null device
set_null_device("png")
```
 

```{r}
df <- readRDS("./data/dataset/df.RDS")
pclustering <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS") 
gwas_hits <- readRDS("./data/GWAS/hits/gwas_mat.RDS")
nn <- readRDS("./data/nearest_neighbor_comparisons/nn_comparisons.RDS")

df <- left_join(df,gwas_hits %>% mutate(isolate_no = rownames(.))) %>% left_join(.,pclustering)
```

# S Figure 4A - Table
```{r}
s_figure_4.A_tb <- gwas_table %>% select(locus_tag,gene,product,resistance_category,sig_tests,sig_models_simple,sig_phenotypes,nn_qc) 
s_figure_4.A_tb$locus_tag <- recode(s_figure_4.A_tb$locus_tag,"AA018"="AA018 blaKPC-containing plasmid","AA552"="AA552 blaKPC-containing plasmid","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36',"KPNIH1_RS18665" = "PFAV in ompK36")
s_figure_4.A_tb$locus_tag <- ifelse(s_figure_4.A_tb$locus_tag =="KPNIH1_RS18665","ompK36 PFAV",s_figure_4.A_tb$locus_tag)
s_figure_4.A_tb$gene <- ifelse(grepl(pattern = 'ompK36',s_figure_4.A_tb$locus_tag) | s_figure_4.A_tb$locus_tag =='KPNIH1_RS18665',"ompK36",s_figure_4.A_tb$gene) %>% {ifelse(is.na(.),"",.)}
s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$locus_tag == "KPNIH1_RS18100","MdtA/MuxA family multidrug efflux RND <br>transporter periplasmic adaptor subunit",s_figure_4.A_tb$product)
s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$gene=='ompK36','porin OmpC',s_figure_4.A_tb$product)
s_figure_4.A_tb$product <- ifelse(grepl(pattern = 'KPC',s_figure_4.A_tb$locus_tag),"blaKPC-associated plasmid cluster",s_figure_4.A_tb$product)
s_figure_4.A_tb$locus_tag <- gsub("blaKPC-","<i>bla<sub>KPC</sub></i>-",s_figure_4.A_tb$locus_tag)
s_figure_4.A_tb$locus_tag <- gsub("ompK36","<i>ompK36</i>",s_figure_4.A_tb$locus_tag) 
s_figure_4.A_tb$gene <- gsub("ompK36","<i>ompK36</i>",s_figure_4.A_tb$gene) 
s_figure_4.A_tb$product <- gsub("blaKPC","<i>bla<sub>KPC</sub></i>",s_figure_4.A_tb$product) 
s_figure_4.A_tb$locus_tag <-  gsub("associated","containing",s_figure_4.A_tb$locus_tag)
s_figure_4.A_tb$product <-  gsub("associated","containing",s_figure_4.A_tb$product)
s_figure_4.A_tb$sig_models_simple <- gsub("KPC-","<i>bla<sub>KPC</sub></i>-",s_figure_4.A_tb$sig_models_simple) %>% gsub("associated","containing",.)
s_figure_4.A_tb$sig_models_simple <- gsub("High-risk variants","Putative function-altering variants",s_figure_4.A_tb$sig_models_simple)

s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$locus_tag == "SHV-11","Broad-spectrum β-lactamase",s_figure_4.A_tb$product)
s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$locus_tag == "TEM-1D.v1^","Broad-spectrum β-lactamase",s_figure_4.A_tb$product)

s_figure_4.A_tbl <- s_figure_4.A_tb %>% select(locus_tag,gene,product,sig_tests,sig_models_simple,sig_phenotypes,nn_qc) %>% `colnames<-`(c('Significant hit','Gene', 'Product', 'Significant tests','Significant models', 'Phenotypes','Nearest neighbor\ncriterion'))

tbl_html <- s_figure_4.A_tbl  %>%
  kable(format = "html", row.names = FALSE,escape = FALSE) %>%
  kable_styling(
    bootstrap_options = c("condensed", "responsive"),
    full_width = TRUE,
    fixed_thead  = TRUE,
    font_size = 44
  ) %>%  row_spec(0, background = "lightgray", color = "black", bold = TRUE)  %>% row_spec(1:8, background = "white") 

save_kable(x = tbl_html, file = "tab_kbl.html")

webshot2::webshot("tab_kbl.html", "tab_kbl.png", vwidth = 3850, vheight = 700) 
```

# S Figure 4B - Phylogeny 
```{r ,fig.height=12.5,fig.width=12.5,echo=F,message=F,error=F,warning=F} 
rownames(df) <- df$isolate_no
s_figure_4.df_mat <- get_gwas_hit_matrix(gwas_table,df)  
sig_hits_name <- recode(colnames(s_figure_4.df_mat),"AA018"= "AA018 blaKPC-containing plasmid","AA552"="AA552 blaKPC-containing plasmid","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36',"KPNIH1_RS18665_3.2" = "PFAV in ompK36")
 
s_figure_4.phylo <- gwas_figures(df,tr,s_figure_4.df_mat,sig_hits_name = sig_hits_name %>% gsub("_1|_2|_1.2|_2.2|_3|_3.2","",.)) + ylim(NA,555)  
s_figure_4.FB <- plot_grid(s_figure_4.phylo,labels = "B",label_size=24)  
```

# S Figure 4C - Resistance Frequency
```{r,fig.height=7.5,fig.width=10}
gwas_nons_freq <- variants_freq(gwas_table$genotype,df)  
gwas_nons_freq$locus_tag <- recode(gwas_nons_freq$locus_tag,"AA018"="AA018 blaKPC-containing plasmid","AA552"="AA552 blaKPC-containing plasmid","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36',"KPNIH1_RS18665" = "PFAV in ompK36")
gwas_nons_freq$`Significant Hit` <- gwas_nons_freq$locus_tag
gwas_nons_freq$`Significant Hit` <- gsub("blaKPC","<i>bla<sub>KPC</sub></i>",gwas_nons_freq$`Significant Hit`)
gwas_nons_freq$`Significant Hit` <- gsub("ompK36","<i>ompK36</i>",gwas_nons_freq$`Significant Hit`)
gwas_nons_freq$`Significant Hit` <- factor(gwas_nons_freq$`Significant Hit`, levels=rev(unique(gwas_nons_freq$`Significant Hit`)))  

s_figure_4.C <- freq_plot(gwas_nons_freq)  +s_figure_4_descriptive_plot_theme
```

# S Figure 4D - Nearest neighbor Data
```{r,fig.height=7.5,fig.width=10} 
nn_data_melt <- lapply(gwas_table$genotype,generate_nn_melt_data,nn_data=nn) %>% do.call(rbind,.) 
nn_data_melt$locus_tag <- recode(nn_data_melt$locus_tag,"AA018"="AA018 blaKPC-containing plasmid","AA552"="AA552 blaKPC-containing plasmid","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36',"KPNIH1_RS18665" = "PFAV in ompK36")
nn_data_melt$locus_tag <-gsub("blaKPC","<i>bla<sub>KPC</sub></i>",nn_data_melt$locus_tag)
nn_data_melt$locus_tag <-gsub("ompK36","<i>ompK36</i>",nn_data_melt$locus_tag)
nn_data_melt$locus_tag <- factor(nn_data_melt$locus_tag, levels=rev(unique(nn_data_melt$locus_tag)))  

s_figure_4.D <- nn_plot(nn_data_melt) + s_figure_4_descriptive_plot_theme
```

# Merge figure
```{r,fig.height=20,fig.width=25}
s_figure_4.FCD <- plot_grid(s_figure_4.C,s_figure_4.D ,nrow = 2,labels = c("C","D"),label_size = 28,align = "hv",label_x = -0.02)
s_figure_4.FCDB <- plot_grid(s_figure_4.FB,s_figure_4.FCD,ncol = 2,rel_heights = c(1,1),rel_widths = c(1,0.625)) 
```

```{r ,fig.height=18.5,fig.width=21.5}
s_figure_4.A <- ggdraw() + draw_image("tab_kbl.png")
unlink(c("tab_kbl.html", "tab_kbl.png"))
s_figure_4.FA <- plot_grid(s_figure_4.A, labels = "A", label_size = 28, align = "hv", rel_widths = c(1))  
s_figure_4 <- plot_grid(s_figure_4.FA,s_figure_4.FCDB,labels = c("",""),nrow=2,label_size = 28, rel_heights = c(0.35,1))   
s_figure_4
```

```{r,message=F,echo=F,error=F,warning=F,fig.width=20,fig.height=12.5} 
ggsave(plot = s_figure_4, filename = './figures/s_figure_4.png', width = 21.5, height = 18.5,limitsize=FALSE,bg='white',dpi=900)
```
