---
title: "Supplemental Figure 8 Differential potential for emergence and spread of β-lactam/β-lactamase inhibitor resistance-associated genotypes "
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r}
packages <- c("tidyverse","cowplot","ggtree","ape","phytools","gridExtra","kableExtra","colorRamp2","RColorBrewer","ComplexHeatmap",'phyloAMR')

# Load packages
invisible(lapply(packages,library,character.only=T,quietly=T))
source("./lib/common_functions.R")
source("./lib/consistent_themes.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R") 

# Set null device
set_null_device("png")
```

```{r} 
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
# GWAS table
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS") %>% subset(nn_qc=="Pass")
gwas_mat <- readRDS("./data/GWAS/hits/gwas_mat.RDS")  
# Matrices
## Kleborate matrix
kleborate <- readRDS("./data/GWAS/matrices/kleborate_mat.RDS") %>% as.data.frame 
## KPC Plasmids 
KPC_containing_clusters <- readRDS("./data/GWAS/matrices/KPC_containing_cluster_mat.RDS") %>% as.data.frame
## Core genome burden models
models <- c("model1","model1.2","model2","model2.2","model3","model3.2")
for(x in models){
  grouped_model <- paste0(x,"_burden.RDS")
  assign(gsub(".RDS","",grouped_model),readRDS(paste0("./data/GWAS/matrices/",grouped_model)) %>% as.data.frame)
}
## BLBLI clustering
pclustering <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS")
blbli_asr <- readRDS("./data/asr_clustering/blbli_asr.RDS")
## Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
# Join
df <- left_join(df,kleborate %>% mutate(isolate_no = rownames(.)))
df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.)))  
df <- left_join(df,pclustering %>% select(isolate_no,blbli_asr_cluster_renamed))
df$blbli_asr_cluster_renamed_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible" )
   
df <- left_join_matrix(model1_burden,df,"_1")
df <- left_join_matrix(model1.2_burden,df,"_1.2")
df <- left_join_matrix(model2_burden,df,"_2")
df <- left_join_matrix(model2.2_burden,df,"_2.2")
df <- left_join_matrix(model3_burden,df,"_3")
df <- left_join_matrix(model3.2_burden,df,"_3.2")
df <- left_join(df,canonical %>% mutate(isolate_no = rownames(.))) 
```

# A - Heatmap for presence by resistance category
```{r,fig.height=7,fig.width=12}
source("./scripts/figures/s_figure_8_functions.R")

# Heatmap curation
df$blbli_asr_cluster_renamed_simple <- ifelse(df$blbli_asr_cluster_renamed_simple=="Singleton" & df$clade_I == "Clade I","Singleton - Clade I",ifelse(df$blbli_asr_cluster_renamed_simple == "Singleton" & df$clade_I == "Clade II","Singleton - Clade II",ifelse(df$blbli_asr_cluster_renamed_simple=="No feature" & df$clade_I == "Clade I","Susceptible - Clade I", ifelse(df$blbli_asr_cluster_renamed_simple=="No feature" & df$clade_I == "Clade II","Susceptible - Clade II",df$blbli_asr_cluster_renamed_simple)))) 

variables <- c(gwas_table$genotype,canonical %>% select(-isolate_no) %>% colnames(.) %>% sort) %>% unique %>% subset(!. %in% c("OmpK35-25%","OmpK36_loop3_insertion","OmpK36_L3_mutations","PBP_any","FJ223607","OmpK36_Loss_of_function","OmpK36_intergenic","OmpK36_non_syn","KPNIH1_RS07805_2.2",'OmpK36_putative_function_altering'))
prop_data <- lapply(variables,numerical_data,df=df) %>% do.call(rbind,.) %>% `rownames<-`(variables) 
prop_columns <- colnames(prop_data) %>% subset(. != "Cluster_prop" & grepl("_prop",.)) %>% sort
ct_columns <- colnames(prop_data) %>% subset(. != "Cluster_ct" & grepl("_ct",.)) %>% sort
df$blbli_asr_cluster_renamed_simple <- factor(df$blbli_asr_cluster_renamed_simple,levels = c(prop_columns %>% gsub("_prop","",.)))
total_ct <-  table(df %>% .$blbli_asr_cluster_renamed_simple)  
total_genotype_ct <- prop_data %>% select(ct_columns) %>% `colnames<-`(gsub("_ct","",colnames(.))) %>% rowSums()
column_names <- prop_columns %>% gsub("_prop","",.) %>% paste0(.," (n=",total_ct,")")
ct_matrix <-  prop_data %>% select(ct_columns) %>% `colnames<-`(gsub("_ct","",colnames(.)))  %>% round(.,0)  %>% as.matrix %>% {ifelse(. ==0,"",.)}
row_names <- variables %>% gsub("_1.2|_1|_2.2|_2|_3|_3.2","",.)  %>% recode(.,'OmpK35-25'='Truncation in ompK35 porin at 25%','OmpK36_c25t'='25 cytosine-to-thymine transition in ompK36','OmpK36_L3_mutations'='Loop 3 insertion in ompK36','OmpK36GD'='GD loop 3 insertion in ompK36','OmpK36TD'='TD loop 3 insertion in ompK36','OmpK36_truncation_kleborate'='Truncation in ompK36', 'OmpK36_non_syn'='Non-synonymous mutation in ompK36','OmpK36_putative_function_altering'='PFAV in ompK36','OmpK36_intergenic'='Mutation in 5\' intergenic region of ompK36','OmpK36_promoter_IS'='Insertion sequence at ompK36 promoter',"AcrAB_TolC_any" = 'acrAB-tolC efflux pump mutant','RamR'='PFAV in ramR efflux pump regulator',"RamA"='PFAV in ramA efflux pump activator','PBP_any'='Non-synonymous mutations in penicillin-binding-proteins','PBP2'='Penicillin-binding protein-2 mutant',"PBP4"='Penicillin-binding protein-4 mutant',"AA018"= "AA018 blaKPC-containing plasmid","AA552"="AA552 blaKPC-containing plasmid","KPNIH1_RS18665" = "PFAV in ompK36")
row_names <- gsub("blaKPC","<i>bla<sub>KPC</sub></i>",row_names)
row_names  <- gsub("ompK36","<i>ompK36</i>",row_names )  
row_names  <- gsub("ompK35","<i>ompK35</i>",row_names )  
row_names  <- gsub("acrAB-tolC","<i>acrAB-tolC</i>",row_names )  
row_names  <- gsub("ramR","<i>ramR</i>",row_names )  
row_names  <- gsub("ramA","<i>ramA</i>",row_names )   
row_names <- row_names %>% paste0(.," (n=",total_genotype_ct,")") 
prop_matrix <- prop_data %>% select(prop_columns) %>% `colnames<-`(gsub("_prop","",colnames(.))) %>% round(.,3) %>% as.matrix 


fix_hyphens <- function(x) {
gsub("-", "\u202F-\u202F", x)
} 
row_names <- fix_hyphens(row_names)
remove_hyphen_spaces <- function(x) gsub("[\u202F\u00A0 ]*-[\u202F\u00A0 ]*", "-", x)
row_names <- remove_hyphen_spaces(row_names)

# Legend
one_hundred_breaks <- c()
for(i in 1:100){
  one_hundred_breaks[i] <- 1*i
}  

fig_legend <- Legend(at = c(0, 0.25, 0.50, 0.75, 1.00),labels = c("0%", "25%", "50%", "75%", "100%"),title = "Proportion",direction = "horizontal",col_fun = colorRamp2(breaks = c(0, 1.00),colors = c("white","red")), title_gp = gpar(fontsize = 16),labels_gp=gpar(fontsize=14),legend_width = unit(12, "cm"))

heatmap_legend_params <- list(
  title = "Percentage",
  at = c(0, 0.25, 0.5, 0.75, 1),
  labels = c("0%", "25%", "50%", "75%", "100%"),
  direction = "horizontal",
  title_gp = gpar(fontsize = 16),
  labels_gp = gpar(fontsize = 14),
  legend_width = unit(10, "cm")
)

CI_groups <- table(df$clade_I,df$blbli_asr_cluster_renamed_simple) %>% .[1,] %>% subset(.>0) %>% names 

column_names_nnotated <- lapply(colnames(prop_matrix),FUN=function(x){
  if (x %in% CI_groups) {
    y <- "Clade I"
      names(y) <-x
  } else {
    y <- "Clade II"
    names(y) <- x 
  }
  y
   
}) %>% unlist

clade_labels <- columnAnnotation(Clade = column_names_nnotated,col=  list(Clade = clade_colors),show_legend=F)

# Heatmap
A_heatmap <- ComplexHeatmap::Heatmap(prop_matrix,show_heatmap_legend = T,show_row_names = T,col = c("white","red"),cluster_columns = T,cluster_rows=T,name = "Percentage",column_labels = column_names,row_labels = gt_render(row_names), cell_fun = function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.3s", ct_matrix[i, j]), x, y, gp = gpar(fontsize = 11))
},border_gp = gpar(col = "black", lty = 1), rect_gp = gpar(col = "white", lwd = 1),bottom_annotation = clade_labels,row_names_gp =  gpar(fontsize = 11),column_names_gp =  gpar(fontsize = 13),heatmap_legend_param = heatmap_legend_params, width = unit(10, "cm"),
 row_names_max_width = unit(8.759, "cm"))
 
# Save heatmap to a temporary PNG file
tmpfile <- tempfile(fileext = ".png")
ht <- draw(A_heatmap, heatmap_legend_side = "bottom") 
dev.new()
png(tmpfile,
    width = 8 * 900,    # 9.9 inches * 300 dpi = 2970 pixels
    height = 7 * 900,   # 7.5 inches * 300 dpi = 2250 pixels
    res = 900)

draw(ht,padding = unit(c(0, -20, 0, 0), "cm"))
dev.off() 
```

# B - Occurrence
```{r}
traits_of_interest <- c("blbli_dich_num",rownames(prop_matrix))   
trait_parent_child <- pbmcapply::pbmclapply(traits_of_interest,FUN=function(x){
  model <- ifelse(x == "blbli_dich_num","ER","MF")
  phyloAMR::asr(df = df,tr = tr,tip_name_variable = 'isolate_no',trait=x,model=model) %>% .$parent_child_df
},mc.cores=7)
names(trait_parent_child) <- traits_of_interest

synchronous_changes_w_blbli <- pbmcapply::pbmclapply(traits_of_interest %>% subset(.!="blbli_dich_num"),FUN=function(x){
  phyloAMR::synchronous_transitions(comparitor_parent_child_df = trait_parent_child[[x]],trait_parent_child_df = trait_parent_child[['blbli_dich_num']],node_states = 'joint')
},mc.cores=7) %>% do.call(rbind.data.frame,.) %>% mutate(comparitor = traits_of_interest %>% subset(.!="blbli_dich_num"))

genotypes_w_transitions_in_parental_nodes <- apply(synchronous_changes_w_blbli,MARGIN = 1,FUN=function(x){
  transitions = x['synchronous_gains'] %>% str_split(.,",") %>% unlist
  
  ifelse(sum(transitions>(tr$Nnode+1))>0,T,F)
})

genotypes <- synchronous_changes_w_blbli[genotypes_w_transitions_in_parental_nodes,'comparitor'] %>% subset(is.na(.)==F)
```


## Data
```{r}
st258 <- ggtree(tr, size=0.25)
genotypes_and_blbli <- c('blbli_dich_num',genotypes,"RamA")

asr_traits_of_interest <-  pbmcapply::pbmclapply(genotypes_and_blbli,FUN=function(x){
  asr_cluster_detection(df = df,tr = tr,'isolate_no',patient_id='Patient_ID',parent_child_df = trait_parent_child[[x]],node_states = 'joint',confidence = NULL,simplify_faux_clusters = T,simplify_revertant = T,collapse_cluster = T)
},mc.cores=7)
names(asr_traits_of_interest) <- genotypes_and_blbli
 
gains <- lapply(genotypes_and_blbli,FUN=function(y){
  asr_traits_of_interest[[y]][['asr_cluster']] %>% subset(grepl("cluster",.)) %>% str_split("_",simplify = T) %>% .[,2] %>% unlist %>% unique %>% as.numeric 
}) %>% `names<-`(genotypes_and_blbli)

tips <- hues::iwanthue(5,plot=T)
 
tree_w_blbli <- st258 + geom_nodepoint(aes(subset=(node %in% gains$blbli_dich_num),fill="BL/BLI" ), shape=21, size= 4,show.legend = T) +  scale_fill_manual(name = "Phylogenetic occurrence", values ="red",breaks = "BL/BLI" )
```
## Create figure
```{r}
blbli_name = "BL/BLI resistance"
AA552_name = "AA552 <i>bla<sub>KPC</sub></i>-containing plasmid"
ompK36_name = "PFAV in <i>ompK36</i>"
PBP_name = "PBP-2 mutant"
RamA_name = "PFAV in <i>ramA</i>" 
genotype_ordering <- c(blbli_name ,AA552_name,ompK36_name,PBP_name,RamA_name)

B <- st258 + 
    geom_nodepoint(aes(fill = blbli_name,shape=blbli_name,subset=(node %in% gains$blbli_dich_num),size=blbli_name,legend=F)) +
    geom_nodepoint(aes(fill = AA552_name,shape=AA552_name,subset=(node %in% gains$AA552),size=AA552_name,legend=F)) +
    geom_nodepoint(aes(fill = ompK36_name,shape=ompK36_name,subset=(node %in% gains$KPNIH1_RS18665_3.2),size=ompK36_name,legend=F)) +  
  geom_nodepoint(aes(fill = RamA_name,shape=RamA_name,subset=(node %in% gains$RamA),size=RamA_name,legend=F))  +
    geom_nodepoint(aes(fill = PBP_name,shape=PBP_name,subset=(node %in% gains$PBP2),size=PBP_name,legend=F)) + 
  scale_fill_manual(name="Synchronous clustering",values = setNames(
    c("red", "#91796A", "#97C466", "#9451AD","#97B2B7"), genotype_ordering
  ),breaks = genotype_ordering,guide = guide_legend(ncol=1,title.position = "top", label.position = "right")) + 
    scale_shape_manual(name="Synchronous clustering",values=setNames(
    c(21,22,23,24,25),genotype_ordering
  ),breaks=genotype_ordering,guide = guide_legend(ncol=1,title.position = "top", label.position = "right")) + 
  scale_size_manual(name="Synchronous clustering",values = setNames(
    c(4,2.25,2.5,1.75,2.25),genotype_ordering),breaks=genotype_ordering,guide = guide_legend(ncol=1,title.position = "top", label.position = "right"))  + 
  scale_alpha(guide = 'none') + theme(legend.position = 'bottom')
rownames(df) <- df$isolate_no

B_1 <- B + ggnewscale::new_scale_fill()

B_w_tree <- gheatmap(B_1,df %>% select(blbli_asr_cluster_renamed) %>% `colnames<-`("BL/BLI clustering"),width=0.05,colnames_position = 'top',colnames_angle = 90, color=NULL ,colnames_offset_y = 0.25, hjust = 0) + cluster_scale_2_col + ylim(NA,560) + theme(legend.position = 'bottom',legend.text = element_markdown())
```

# Cowplot supplemental figure
```{r, fig.height=6.5, fig.width=11.375} 
s_figure_8.A <- ggdraw() + draw_image(tmpfile)
unlink(tmpfile)
s_figure_8 <- cowplot::plot_grid(s_figure_8.A,B_w_tree,labels = "AUTO",ncol=2,label_size = 20,rel_widths = c(1,0.625))
s_figure_8
``` 

```{r,message=F,echo=F,error=F,warning=F,fig.width=20,fig.height=12.5} 
ggsave(plot = s_figure_8, filename = './figures/s_figure_8.png', width = 11.375, height = 6.5,limitsize=FALSE,bg='white',dpi=900)
```