---
title: "s_figure_8_alignment_tn4401"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r , echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse","gridExtra",'kableExtra',"ggnewscale","cowplot",'ape','ggtree')

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts  
source("./lib/common_functions.R")   
source("./lib/consistent_themes.R") 
```

```{r}
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile") 

# Matrices 
## KPC Plasmids 
KPC_info_clusters <- readRDS("./data/KPC_plasmid/KPC_containing_clusters_mat.RDS") %>% as.data.frame 

KPC_containing_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_KPC")) %>% {ifelse(.=="KPC Plasmid",1,0)} %>% as.data.frame%>% select_if(colSums(.)>0)

KPC_clusters <- colnames(KPC_containing_clusters)

Non_KPC_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_non_KPC")) %>% {ifelse(.=="Non-KPC plasmid",1,0)} %>% as.data.frame %>% select_if(colSums(.)>0)

df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.))) %>% left_join(Non_KPC_clusters %>% mutate(isolate_no = rownames(.)))
```

```{r}
df <- df %>% mutate(reason_for_interest = ifelse(isolate_no =="PCMP_H339","Other AA552 Plasmid: LS998787",
                                 ifelse(isolate_no  %in% c("PCMP_H148"),"Other AA552 Plasmid: CP032203",
                                 ifelse(isolate_no %in% c("PCMP_H434"),"Clade IIA FJ223607 Plasmid",
                                        ifelse(isolate_no %in% c("PCMP_H30","PCMP_H301","PCMP_H307"),"Non-Susceptibility Cluster 5",
                                               ifelse(isolate_no %in% c("PCMP_H416","PCMP_H15"),"Non-Susceptibility Cluster 6 Member",
                                                      ifelse(isolate_no %in% c("PCMP_H173","PCMP_H276"),"Resistance Singleton Pair","Not Interested")))
                                              ))))
 
```

```{r}
KPC_type <- c()
for(i in 1:nrow(df)){
  rows <- df %>% select(KPC_clusters) %>% .[i,]  
  KPC_plasmid <- which(rows ==1)
  if(length(KPC_plasmid)>0){
  KPC_type[[i]] <- rows %>% select(KPC_plasmid) %>% colnames() 
  } else {
  KPC_type[[i]] <- "No KPC Plasmid"
  }  
}

KPC_type <- KPC_type %>% unlist
df$KPC_plasmid <- KPC_type %>% gsub("_KPC","",.) %>% as.factor
df$KPC_plasmid <- relevel(df$KPC_plasmid,ref = "No KPC Plasmid")
```

# Long read data

```{r}
blast_lr <- read.delim("./data/KPC_blast_ONT/KPC_blast_master_ONT.txt",sep = ' ') #%>% subset
blast_lr$contig <- blast_lr$qseqid %>% str_split(.,";",simplify = T) %>% .[,1] %>% unlist %>% str_split(.,"_",simplify=T) %>% .[,3:4]  %>% as.data.frame %>% apply(.,1,paste0,collapse="_")
blast_lr$element_symbol <- ifelse(grepl("KPC-2",blast_lr$sseqid),"KPC-2",ifelse(grepl("KPC-3",blast_lr$sseqid),"KPC-3",ifelse(grepl("KPC-5",blast_lr$sseqid),"KPC-5","Other")))
blast_lr$gene <- blast_lr$element_symbol

mobtyper_lr <- read.delim("./data/mobsuite_ONT/mobtyper_results_ONT.txt")  
mobcontigs_lr <- read.delim("./data/mobsuite_ONT/mobsuite_contig_report_ONT.txt")
mobcontigs_lr$contig <- mobcontigs_lr$contig_id%>% str_split(.,";",simplify = T) %>% .[,1] %>% unlist %>% str_split(.,"_",simplify=T) %>% .[,3:4]  %>% as.data.frame %>% apply(.,1,paste0,collapse="_")
mobcontigs_lr$isolate_no <- gsub("_medaka","",mobcontigs_lr$sample_id)

KPC_data <- left_join(blast_lr,mobcontigs_lr)
lr_KPC_cluster <-  distinct(KPC_data %>% select(isolate_no,element_symbol,primary_cluster_id,mash_nearest_neighbor)) %>% `colnames<-`(c("isolate_no","KPC gene","KPC_cluster_lr","KPC_plasmid_lr"))
lr_isolates <- subset(df,reason_for_interest != "Not Interested")
lr_isolates <- left_join(lr_isolates,lr_KPC_cluster)
```

# Read in data

```{r}
kpc_contigs <- blast_lr %>% subset(gene  %in% c('KPC',"KPC-2","KPC-3","KPC-5")) %>% select(isolate_no,contig,gene) %>% distinct

contig_data <- mobcontigs_lr %>% select(isolate_no,contig,primary_cluster_id,mash_nearest_neighbor,size) %>% left_join(kpc_contigs)
contig_data$primary_cluster_id <- ifelse(contig_data$primary_cluster_id == "-","Chromosome",contig_data$primary_cluster_id) 
contig_data$contig_name <- paste0(contig_data$contig,"-",contig_data$primary_cluster_id,"-",contig_data$mash_nearest_neighbor,"-",contig_data$gene,"-",contig_data$size,"bp") %>% gsub("---","",.) %>% gsub("NA","",.) %>% trimws(.,"right","-") %>% gsub("--","-",.)
```

# tn4401 data

```{r}
tn4401_blast <- read.delim("./data/tn4401_blast_ONT/tn4401_blast_ONT_master.txt",sep = "")
tn4401_blast$isolate_no <- gsub("_lr","",tn4401_blast$isolate_no )
tn4401_blast$qseqid <- str_split(tn4401_blast$qseqid,";",simplify = T) %>% .[,1] %>% unlist %>% str_split(.,"_",simplify=T) %>% .[,3:4]  %>% as.data.frame %>% apply(.,1,paste0,collapse="_")
```

# Create figures

```{r ,fig.width=12.5,fig.height=10}
library(tidyverse)
library(ggplot2)

 reorder_blast_df <- function(blast_isolate_df,name,contig_name,num){ 
  if(blast_isolate_df["sstart"]>blast_isolate_df["send"]){
    blast_df <- blast_isolate_df %>% t %>% as.data.frame %>% `colnames<-`(c(name,contig_name,paste0("[E",num,"]"),paste0("[S",num,"]"),"send","sstart"))
  } else {
    
    blast_df <- blast_isolate_df %>% t %>% as.data.frame %>% `colnames<-`(c(name,contig_name,paste0("[S",num,"]"),paste0("[E",num,"]"),"sstart","send")) 
  } 
  return(blast_df)
 }
 
 

make_plots <- function(filename,contig_data,blast_data){
  coords <- read_delim(filename, 
                       delim = "\t", escape_double = FALSE, 
                       trim_ws = TRUE, skip = 3,col_names = T) %>% as.data.frame
  coords$`[TAGS]` <- gsub("\\t"," ",coords$`[TAGS]`)
  coords$map1_chr <- str_split(coords$`[TAGS]`," ",simplify=T) %>% .[,1] %>% str_split(.,";",simplify = T) %>% .[,1] %>% unlist %>% str_split(.,"_",simplify=T) %>% .[,3:4]  %>% as.data.frame %>% apply(.,1,paste0,collapse="_")
  coords$map2_chr <- str_split(coords$`[TAGS]`," ",simplify=T) %>% .[,2] %>% str_split(.,";",simplify = T) %>% .[,1] %>% unlist %>% str_split(.,"_",simplify=T) %>% .[,3:4]  %>% as.data.frame %>% apply(.,1,paste0,collapse="_")
  coords$map1_name <- gsub(".coords","",filename) %>% str_split(.,"_",simplify=T) %>% .[,1:2] %>% paste0(collapse="_")
  coords$map2_name <- gsub(".coords","",filename) %>% str_split(.,"_",simplify=T) %>% .[,3:4] %>% paste0(collapse="_")
  coords$map1_pos <- coords$`[S1]`
  coords$map2_pos <- coords$`[S2]` 
  
  # Contig renaming
  ## Isolate 1
  isolate_1 <- coords$map1_name %>% unique
  isolate_1_contigs <- contig_data %>% subset(isolate_no == isolate_1)
  isolate_1_contigs_names <- isolate_1_contigs$contig_name  %>% `names<-`(isolate_1_contigs$contig)
  coords$map1_chr <- recode(coords$map1_chr,!!!as.list(isolate_1_contigs_names))
  coords$map1_chr <- factor(coords$map1_chr,levels = c(isolate_1_contigs %>% arrange(-size) %>% .$contig_name))
  ### tn4401 data
  tn4401_blast_1 <- blast_data %>% subset(isolate_no == isolate_1)%>% select(isolate_no,qseqid,qstart,qend,sstart,send) 
  tn4401_blast_1 <- apply(tn4401_blast_1,1,FUN=reorder_blast_df,name="map1_name",contig_name="map1_chr",num="1") %>% bind_rows
  tn4401_blast_1$map1_chr <- recode(tn4401_blast_1$map1_chr,!!!as.list(isolate_1_contigs_names))
  tn4401_blast_1$map1_chr <- factor(tn4401_blast_1$map1_chr,levels = c(isolate_1_contigs %>% arrange(-size) %>% .$contig_name)) 
  ## Isolate 2
  isolate_2 <- coords$map2_name %>% unique
  isolate_2_contigs <- contig_data %>% subset(isolate_no == isolate_2)
  isolate_2_contigs_names <- isolate_2_contigs$contig_name  %>% `names<-`(isolate_2_contigs$contig)
  coords$map2_chr <- recode(coords$map2_chr,!!!as.list(isolate_2_contigs_names))
  coords$map2_chr <- factor(coords$map2_chr,levels = c(isolate_2_contigs %>% arrange(-size) %>% .$contig_name))
  tn4401_blast_2 <- blast_data %>% subset(isolate_no == isolate_2)  %>% select(isolate_no,qseqid,qstart,qend,sstart,send) 
  tn4401_blast_2 <- apply(tn4401_blast_2,1,FUN=reorder_blast_df,name="map2_name",contig_name="map2_chr",num="2") %>% bind_rows
  tn4401_blast_2$map2_chr <- recode(tn4401_blast_2$map2_chr,!!!as.list(isolate_2_contigs_names))
  tn4401_blast_2$map2_chr <- factor(tn4401_blast_2$map2_chr,levels = c(isolate_2_contigs %>% arrange(-size) %>% .$contig_name)) 
  # Merge KPC data 
  tn4401_merge <- left_join(tn4401_blast_1,tn4401_blast_2)
  tn4401_merge <- tn4401_merge %>% mutate_if(colnames(.) %in% c("[E1]",'[E2]','[S1]','[S2]'),as.numeric) 
  # Contig barriers
  contig_data <- lapply(isolate_1_contigs_names,FUN=function(x){
    contig_info_1 <- isolate_1_contigs %>% subset(contig_name==x)  %>% select(contig_name,size) %>% `colnames<-`(c("map1_chr","X"))
  combinded <- lapply(isolate_2_contigs_names,function(y){
      contig_info_2 <- isolate_2_contigs %>% subset(contig_name==y)  %>% select(contig_name,size) %>% `colnames<-`(c("map2_chr","Y"))
      cbind(contig_info_1,contig_info_2)
    }) %>% do.call(rbind,.) %>% `rownames<-`(NULL)
    return(combinded)
  }) %>% do.call(rbind,.) 
  contig_data$map1_chr <- factor(contig_data$map1_chr,levels = c(isolate_1_contigs %>% arrange(-size) %>% .$contig_name)) 
  contig_data$map2_chr <- factor(contig_data$map2_chr,levels = c(isolate_2_contigs %>% arrange(-size) %>% .$contig_name)) 
  
    figure <- ggplot(data=coords, aes(x=`[S1]`, xend=`[E1]`, y=`[S2]`, yend=`[E2]`)) + geom_segment() +  geom_segment(data=tn4401_merge,aes(x=`[S1]`, xend=`[E1]`, y=`[S2]`, yend=`[E2]`,color="red"))  + geom_segment(data=contig_data,aes(x=`X`,xend=`X`,y=`Y`,yend=`Y`),color="white") +
    facet_grid(map2_chr~map1_chr,scales = 'free',drop = F)   + xlab(paste0(unique(coords$map1_name)," position (bp)")) + ylab(paste0(unique(coords$map2_name)," position (bp)")) +    geom_point(alpha=.5)  + xlim(0,NA) + ylim(0,NA)+ coord_cartesian(clip = "off") + ggplot2::theme_bw()  + scale_color_manual(values="red",labels="tn4401 KPC transposon", name="", guide = guide_legend(nrow=1, title.position = "top", label.position = "right")) + theme(legend.position = "bottom") + theme(legend.position = "bottom",
                     axis.text =   element_text(size=14,color="black"),
                     axis.title = element_text(size = 18,color="black"),
                     legend.text =   element_text(size=18,color="black"),
                     legend.title = element_text(size = 0,color="black"),
                      strip.text = element_text(size = 12),
                     legend.margin = margin(t = -5, r = 0, b = 0, l = 0),
                     panel.spacing = unit(0.3, "cm", data = NULL)) 
 
  figure
}

```

# Get nearest neighbor of KPC-containing strains

```{r}
lr_isolates <- unique(blast_lr$isolate_no) %>% sort
lr_isolates_CI <- lr_isolates %>% subset(. %in% (df %>% subset(clade_I=='Clade I') %>% .$isolate_no))
lr_isolates_CII <- lr_isolates %>% subset(. %in% (df %>% subset(clade_I!='Clade I') %>% .$isolate_no))
paste0("Plasmid profile")

df$lr_isolates <- ifelse(df$isolate_no %in% lr_isolates,df$isolate_no,"Not sequenced")
rownames(df) <- df$isolate_no



po <- gheatmap(ggtree(tr %>% keep.tip(., df %>% subset(clade_I=='Clade I') %>% .$isolate_no)),df %>% select(blbli_dich_num,AA552_KPC,AA552_non_KPC,AA018_KPC),color=NULL,width = 0.8,colnames_position = 'top',colnames_angle = 90, colnames_offset_y = 0.25, hjust = 0,low='white',high='black')
p0.1 <- po+ggnewscale::new_scale_fill()

gheatmap(p0.1,df %>% select(lr_isolates),color=NULL,offset = 0.00025,width=0.2,colnames_position='top',colnames_angle = 90, colnames_offset_y = 0.25, hjust = 0) + ylim(NA,200) + scale_fill_manual(values = c("white",hues::iwanthue(7)))

lapply(lr_isolates_CI,FUN=function(x){ 
  contig_data_df <- contig_data %>% subset(isolate_no ==x)  
  AA552 <- ifelse(sum(contig_data_df$primary_cluster_id %in% "AA552"),"Yes","No")
  AA018 <- ifelse(sum(contig_data_df$primary_cluster_id %in% "AA018"),"Yes","No")
  KPC_plasmid <- subset(contig_data_df,grepl("KPC",gene)) %>% .$primary_cluster_id
  data.frame(isolate_no=x,AA018,AA552,KPC_plasmid)
}) %>% do.call(rbind,.)


phylo_dist <- readRDS("./data/tree/phylo_dist.RDS") %>% as.data.frame %>% subset(rownames(.) %in% df$isolate_no) %>% select(any_of(df$isolate_no)) %>% as.matrix 

variant_dist <- readRDS("./data/nearest_neighbor_comparisons/pairwise_snp_dist_mat.RDS")  

phyloAMR::nearest_neighbor_algorithm("PCMP_H276",phylogenetic_distance = phylo_dist,variant_distance = variant_dist,metadata = df %>% subset(isolate_no %in% c(lr_isolates)),annotate = F,comparison = T,comparison_feature = 'AA018_KPC')

phyloAMR::nearest_neighbor_algorithm("PCMP_H148",phylogenetic_distance = phylo_dist,variant_distance = variant_dist,metadata = df %>% subset(isolate_no %in% c(lr_isolates)),annotate = F,comparison = T,comparison_feature = 'AA018_KPC')
 
```

# Main figure

```{r,fig.width=13.5,fig.height=11.5}
setwd("./data/mummer_ONT/")
tn4401_hopping <- make_plots(filename = 'PCMP_H276_PCMP_H173.coords',contig_data = contig_data,blast_data = tn4401_blast) 
tn4401_hopping

ggsave("../../figures/s_figure_8.png",plot=tn4401_hopping,width = 13.5,height = 11.5)
```

# Alternative

```{r,fig.width=18.5,fig.height=18.5}
setwd("./data/mummer_ONT/")
tn4401_hopping_2 <- make_plots(filename = 'PCMP_H148_PCMP_H307.coords',contig_data = contig_data,blast_data = tn4401_blast) 
tn4401_hopping_2
```