---
title: "Supplemental figure 4: All GWAS hits"
output: html_document
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
---

# Combination Therapy Manuscript
```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

# Environment
```{r}
packages <- c("tidyverse","cowplot","ggtree","ggnewscale","ape","phytools","phyloAMR","gridExtra")

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./lib/common_functions.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")

# Print
Sys.info()
sessionInfo() 
```
 

```{r}
df <- readRDS("./data/dataset/df.RDS")
pclustering <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS") 
gwas_hits <- readRDS("./data/GWAS/hits/gwas_mat.RDS")
nn <- readRDS("./data/nearest_neighbor_comparisons/nn_comparisons.RDS")

df <- left_join(df,gwas_hits %>% mutate(isolate_no = rownames(.))) %>% left_join(.,pclustering)
```

# S Figure 4A - Table
```{r}
s_figure_4.A_tb <- gwas_table %>% select(locus_tag,gene,product,resistance_category,sig_tests,sig_models_simple,sig_phenotypes)
s_figure_4.A_tb$locus_tag <- recode(s_figure_4.A_tb$locus_tag,"AA018"="AA018 KPC plasmid cluster","AA552"="AA552 KPC plasmid cluster","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36')
s_figure_4.A_tb$gene <- ifelse(grepl(pattern = 'ompK36',s_figure_4.A_tb$locus_tag) | s_figure_4.A_tb$locus_tag =='KPNIH1_RS18665',"ompK36",s_figure_4.A_tb$gene) %>% {ifelse(is.na(.),"",.)}
s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$locus_tag == "KPNIH1_RS18100","MdtA/MuxA family multidrug efflux RND \ntransporter periplasmic adaptor subunit",s_figure_4.A_tb$product)
s_figure_4.A_tb$product <- ifelse(s_figure_4.A_tb$gene=='ompK36','porin OmpC',s_figure_4.A_tb$product)
s_figure_4.A_tb$product <- ifelse(grepl(pattern = 'KPC',s_figure_4.A_tb$locus_tag),"blaKPC-associated plasmid",s_figure_4.A_tb$product)

s_figure_4.A_tbl <- s_figure_4.A_tb %>% select(locus_tag,gene,product,resistance_category,sig_tests,sig_models_simple,sig_phenotypes) %>% `colnames<-`(c('Significant Hit','Gene Hit', 'Product','Genotype Category', 'Significant Tests','Significant Models', 'Significant Phenotypes'))

s_figure_4.A <- s_figure_4.A_tbl %>% tableGrob(rows = NULL,theme=mytheme_GWAS)  
```

# S Figure 4B - Phylogeny 
```{r ,fig.height=12.5,fig.width=12.5,echo=F,message=F,error=F,warning=F} 
rownames(df) <- df$isolate_no
s_figure_4.df_mat <- get_gwas_hit_matrix(gwas_table,df) 
sig_hits_name <- recode(colnames(s_figure_4.df_mat),"AA018"="AA018 KPC plasmid cluster","AA552"="AA552 KPC plasmid cluster","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36')
s_figure_4.phylo <- gwas_figures(df,tr,s_figure_4.df_mat,sig_hits_name = sig_hits_name %>% gsub("_1|_2|_1.2|_2.2|_3|_3.2","",.)) + ylim(NA,545)  
s_figure_4.FB <- plot_grid(s_figure_4.phylo,labels = "B",label_size=24)  
s_figure_4.FB
```

# S Figure 4C - Resistance Frequency
```{r,fig.height=7.5,fig.width=10}
gwas_nons_freq <- variants_freq(gwas_table$genotype,df)  
gwas_nons_freq$locus_tag <- recode(gwas_nons_freq$locus_tag,"AA018"="AA018 KPC plasmid cluster","AA552"="AA552 KPC plasmid cluster","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36')
gwas_nons_freq$`Significant Hit` <- gwas_nons_freq$locus_tag
gwas_nons_freq$`Significant Hit` <- factor(gwas_nons_freq$`Significant Hit`, levels=rev(unique(s_figure_4.A_tb$locus_tag)))  

s_figure_4_descriptive_plot_theme <-  theme(legend.position="bottom",axis.text = element_text(size=19,colour = "black"),axis.title = element_text(size=21,colour = "black"),legend.title = element_text(size=21,colour = "black"),legend.text = element_text(size=19,colour = "black"),legend.title.align=0.5 ,legend.key.size = unit(0.5, "cm"),legend.key.width = unit(0.5, "cm")) 

s_figure_4.C <- freq_plot(gwas_nons_freq)  +s_figure_4_descriptive_plot_theme
s_figure_4.C
```

# S Figure 4D - Nearest neighbor Data
```{r,fig.height=7.5,fig.width=10}
# D. FC MIC (MVB * IR overlay) 
nn_data_melt <- lapply(gwas_table$genotype,generate_nn_melt_data,nn_data=nn) %>% do.call(rbind,.)
nn_data_melt$locus_tag <- recode(nn_data_melt$locus_tag,"AA018"="AA018 KPC plasmid cluster","AA552"="AA552 KPC plasmid cluster","OmpK36_loop3_insertion"="Loop 3 insertion in ompK36","OmpK36GD"="GD loop 3 insertion in ompK36",'OmpK36_c25t' = 'C25T SNP in ompK36','OmpK36_truncation_kleborate'='Truncation of ompK36')
nn_data_melt$locus_tag <- factor(nn_data_melt$locus_tag, levels=rev(unique(s_figure_4.A_tb$locus_tag)))  
      
s_figure_4.D <- nn_plot(nn_data_melt) + s_figure_4_descriptive_plot_theme
s_figure_4.D
```

# Merge figure
```{r,fig.height=20,fig.width=25}
s_figure_4.FCD <- plot_grid(s_figure_4.C,s_figure_4.D ,nrow = 2,labels = c("C","D"),label_size = 28,align = "hv",label_x = -0.02)
s_figure_4.FCDB <- plot_grid(s_figure_4.FB,s_figure_4.FCD,ncol = 2,rel_heights = c(1,1),rel_widths = c(1,0.625)) 
```

```{r ,fig.height=17.25,fig.width=26}
s_figure_4.FA <- plot_grid(s_figure_4.A,labels = "A",label_size=28,align = "hv")  
s_figure_4 <- plot_grid(s_figure_4.FA,s_figure_4.FCDB,labels = c("",""),nrow=2,label_size = 28, rel_heights = c(0.35,1))   
s_figure_4
```

```{r,message=F,echo=F,error=F,warning=F,fig.width=20,fig.height=12.5} 
ggsave(plot = s_figure_4, filename = './figures/s_figure_4.png', width = 26, height = 17.25,limitsize=FALSE)
```
