---
title: "s_figure_10"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r}
packages <- c("tidyverse","cowplot","ggtree","ape","phytools","gridExtra","kableExtra","colorRamp2","RColorBrewer","ComplexHeatmap",'phyloAMR')

# Load packages
invisible(lapply(packages,library,character.only=T,quietly=T))
source("./lib/common_functions.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R") 
```

```{r} 
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
# GWAS table
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS") %>% subset(nn_qc=="Yes")
gwas_mat <- readRDS("./data/GWAS/hits/gwas_mat.RDS")  
# Matrices
## Kleborate matrix
kleborate <- readRDS("./data/GWAS/matrices/kleborate_mat.RDS") %>% as.data.frame 
## KPC Plasmids 
KPC_containing_clusters <- readRDS("./data/GWAS/matrices/KPC_containing_cluster_mat.RDS") %>% as.data.frame
## Core genome burden models
models <- c("model1","model1.2","model2","model2.2","model3","model3.2")
for(x in models){
  grouped_model <- paste0(x,"_burden.RDS")
  assign(gsub(".RDS","",grouped_model),readRDS(paste0("./data/GWAS/matrices/",grouped_model)) %>% as.data.frame)
}
## BLBLI clustering
pclustering <- readRDS("./data/asr_clustering/blbli_asr_clustering_df.RDS")
blbli_asr <- readRDS("./data/asr_clustering/blbli_asr.RDS")
## Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
# Join
df <- left_join(df,kleborate %>% mutate(isolate_no = rownames(.)))
df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.)))  
df <- left_join(df,pclustering %>% select(isolate_no,blbli_asr_cluster_renamed))
df$blbli_asr_cluster_renamed_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible" )
   
df <- left_join_matrix(model1_burden,df,"_1")
df <- left_join_matrix(model1.2_burden,df,"_1.2")
df <- left_join_matrix(model2_burden,df,"_2")
df <- left_join_matrix(model2.2_burden,df,"_2.2")
df <- left_join_matrix(model3_burden,df,"_3")
df <- left_join_matrix(model3.2_burden,df,"_3.2")
df <- left_join(df,canonical %>% mutate(isolate_no = rownames(.))) 
```

# A - Heatmap for presence by resistance category
```{r,fig.height=7.5,fig.width=8.75}
source("./scripts/figures/s_figure_10_functions.R")

# Heatmap curation
variables <- c(gwas_table$genotype,canonical %>% select(-isolate_no) %>% colnames(.) %>% sort) %>% unique %>% subset(!. %in% c("OmpK35-25%","OmpK36_loop3_insertion","OmpK36_L3_mutations","PBP_any","FJ223607","OmpK36_Loss_of_function","OmpK36_intergenic","OmpK36_non_syn","KPNIH1_RS07805_2.2",'OmpK36_putative_function_altering'))
prop_data <- lapply(variables,numerical_data,df=df) %>% do.call(rbind,.) %>% `rownames<-`(variables) 
prop_columns <- colnames(prop_data) %>% subset(. != "Cluster_prop" & grepl("_prop",.)) %>% sort
ct_columns <- colnames(prop_data) %>% subset(. != "Cluster_ct" & grepl("_ct",.)) %>% sort
df$blbli_asr_cluster_renamed <- factor(df$blbli_asr_cluster_renamed_simple,levels = c(prop_columns %>% gsub("_prop","",.)))
total_ct <-  table(df %>% .$blbli_asr_cluster_renamed)  
total_genotype_ct <- prop_data %>% select(ct_columns) %>% `colnames<-`(gsub("_ct","",colnames(.))) %>% rowSums()
column_names <- prop_columns %>% gsub("_prop","",.) %>% paste0(.," (n=",total_ct,")")
ct_matrix <-  prop_data %>% select(ct_columns) %>% `colnames<-`(gsub("_ct","",colnames(.)))  %>% round(.,0)  %>% as.matrix %>% {ifelse(. ==0,"",.)}
row_names <- variables %>% gsub("_1.2|_1|_2.2|_2|_3|_3.2","",.)   %>% recode(.,'OmpK36_truncation_kleborate' = 'Truncation in ompK36','OmpK36_c25t'='C25T mutation in ompK36',"AA552"="AA552 KPC plasmid","AA018"="AA018 KPC plasmid","PBP2"="PBP2 modification","PBP4"="PBP4 modification","KPNIH1_RS18665"="PFAV in ompK36","OmpK36_promoter_IS"="IS in promoter of ompK36","RamR"="PFAV in ramR","RamA"="PFAV in ramA","AcrAB_TolC_any"="Variant in AcrAB-TolC efflux pump") %>% paste0(.," (n=",total_genotype_ct,")") 
prop_matrix <- prop_data %>% select(prop_columns) %>% `colnames<-`(gsub("_prop","",colnames(.))) %>% round(.,3) %>% as.matrix 

# Legend
one_hundred_breaks <- c()
for(i in 1:100){
  one_hundred_breaks[i] <- 1*i
}  

fig_legend <- Legend(at = c(0, 0.25, 0.50, 0.75, 1.00),labels = c("0%", "25%", "50%", "75%", "100%"),title = "Proportion",direction = "horizontal",col_fun = colorRamp2(breaks = c(0, 1.00),colors = c("white","red")), title_gp = gpar(fontsize = 16),labels_gp=gpar(fontsize=14),legend_width = unit(12, "cm"))

# Heatmap
A_heatmap <- ComplexHeatmap::Heatmap(prop_matrix,show_heatmap_legend = F,show_row_names = T,col = c("white","red"),cluster_columns = F,cluster_rows=T,name = "Proportion",column_labels = column_names,row_labels = row_names,  cell_fun = function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.3s", ct_matrix[i, j]), x, y, gp = gpar(fontsize = 8))
},border_gp = gpar(col = "black", lty = 1), rect_gp = gpar(col = "white", lwd = 1)) 


fig_heatmap_image <- grid.grabExpr(draw(A_heatmap))
fig_legend_image <- grid.grabExpr(draw(fig_legend))

A <-plot_grid(fig_heatmap_image,NULL,fig_legend_image,ncol=1,axis="l", align="hv",rel_heights = c(1,-.085,.2))   
```

# B - Occurrence
```{r}
traits_of_interest <- c("blbli_dich_num",rownames(prop_matrix))   
trait_parent_child <- pbmcapply::pbmclapply(traits_of_interest,FUN=function(x){
  phyloAMR::asr(df = df,tr = tr,tip_name_var = 'isolate_no',trait=x,model='MF') %>% .$parent_child_df
},mc.cores=5)
names(trait_parent_child) <- traits_of_interest

synchronous_changes_w_blbli <- pbmcapply::pbmclapply(traits_of_interest %>% subset(.!="blbli_dich_num"),FUN=function(x){
  phyloAMR::synchronous_transitions(comparitor_parent_child_df = trait_parent_child[[x]],trait_parent_child_df = trait_parent_child[['blbli_dich_num']],node_states = 'joint')
},mc.cores=5) %>% do.call(rbind.data.frame,.) %>% mutate(comparitor = traits_of_interest %>% subset(.!="blbli_dich_num"))

genotypes_w_transitions_in_parental_nodes <- apply(synchronous_changes_w_blbli,MARGIN = 1,FUN=function(x){
  transitions = x['synchronous_gains'] %>% str_split(.,",") %>% unlist
  
  ifelse(sum(transitions>(tr$Nnode+1))>0,T,F)
})

genotypes <- synchronous_changes_w_blbli[genotypes_w_transitions_in_parental_nodes,'comparitor'] %>% subset(is.na(.)==F)
```


## Data
```{r}
st258 <- ggtree(tr)
genotypes_and_blbli <- c('blbli_dich_num',genotypes)

asr_traits_of_interest <-  pbmcapply::pbmclapply(genotypes_and_blbli,FUN=function(x){
  asr_cluster_detection(df = df,tr = tr,'isolate_no',patient_id='Patient_ID',parent_child_df = trait_parent_child[[x]],node_states = 'joint',confidence = NULL,simplify_faux_clusters = T,simplify_revertant = T,collapse_cluster = T)
},mc.cores=5)
names(asr_traits_of_interest) <- genotypes_and_blbli
 
gains <- lapply(genotypes_and_blbli,FUN=function(y){
  asr_traits_of_interest[[y]][['asr_cluster']] %>% subset(grepl("cluster",.)) %>% str_split("_",simplify = T) %>% .[,2] %>% unlist %>% unique %>% as.numeric 
}) %>% `names<-`(genotypes_and_blbli)

tips <- hues::iwanthue(4,plot=T)
 
tree_w_blbli <- st258 + geom_nodepoint(aes(subset=(node %in% gains$blbli_dich_num),fill="BL/BLI" ), shape=21, size= 4,show.legend = T) +  scale_fill_manual(name = "Phylogenetic occurrence", values ="red",breaks = "BL/BLI" )

genotypes <- gains %>% names %>% .[-1] %>% `names<-`(hues::iwanthue(3,hmin = 15,hmax=360,lmin = 10,plot = T,random = F)  )
```
## Create figure
```{r}
blbli_name = "BL/BLI resistance"
AA552_name = "AA552 KPC plasmid"
ompK36_name = "ompK36 deleterious mutant"
PBP_name = "PBP-2 non-synonymous mutant"
B <- st258 + 
    geom_nodepoint(aes(fill = blbli_name,shape=blbli_name,subset=(node %in% gains$blbli_dich_num),size=blbli_name,legend=F)) +
    geom_nodepoint(aes(fill = AA552_name,shape=AA552_name,subset=(node %in% gains$AA552),size=AA552_name,legend=F)) +
    geom_nodepoint(aes(fill = ompK36_name,shape=ompK36_name,subset=(node %in% gains$KPNIH1_RS18665_3.2),size=ompK36_name,legend=F)) +
    geom_nodepoint(aes(fill = PBP_name,shape=PBP_name,subset=(node %in% gains$PBP2),size=PBP_name,legend=F)) + scale_fill_manual(name="Synchronous clustering",values = setNames(
    c("#91796A", "red", "#97C466", "#9451AD"),
    c(blbli_name, AA552_name, ompK36_name, PBP_name)
  ),labels=c(blbli_name ,AA552_name,ompK36_name,PBP_name),guide = guide_legend(ncol=2,title.position = "top", label.position = "right")) + 
    scale_shape_manual(name="Synchronous clustering",values=setNames(
    c(22,21,23,24),
    c(blbli_name, AA552_name, ompK36_name, PBP_name)
  ),labels=c(blbli_name ,AA552_name,ompK36_name,PBP_name),guide = guide_legend(ncol=2,title.position = "top", label.position = "right")) + scale_size_manual(name="Synchronous clustering",values = setNames(
    c(4,3,3,2),
    c(blbli_name, AA552_name, ompK36_name, PBP_name)),labels=c(blbli_name ,AA552_name,ompK36_name,PBP_name),guide = guide_legend(ncol=2,title.position = "top", label.position = "right"))  + scale_alpha(guide = 'none') + theme(legend.position = 'bottom')
B
```
# Cowplot supplemental figure
```{r,fig.height=7.5,fig.width=14}
A_grob <- ggplotify::as.ggplot(A) + theme(plot.margin = unit(c(0, 0.5, 0, 0.45), 
                                "inches"))
s_figure_10 <- cowplot::plot_grid(A_grob,B,labels = "AUTO",ncol=2,label_size = 20,rel_widths = c(1,.7))
s_figure_10
``` 

```{r,message=F,echo=F,error=F,warning=F,fig.width=20,fig.height=12.5} 
ggsave(plot = s_figure_10, filename = './figures/s_figure_10.png', width = 14, height = 7.5,limitsize=FALSE,bg='white')
```