---
title: "Curate variant calling data"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = TRUE,warning = FALSE, message = FALSE) 
rm(list = ls())
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

# Generate Environment
```{r Generate Environment, results=FALSE,message=FALSE,warning=FALSE}
#Base R Packages
packages <- c("BiocManager","genbankr","snitkitr","tidyverse","ape","phytools","IRanges","rentrez","data.table")
# Function to Load Packages
invisible(lapply(packages,library,character.only=T))

#Functions 
source("./lib/variant_calling_cleaning_functions.R") 

# Print environment
Sys.info()
sessionInfo()
``` 


# Load general data
```{r}
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
```

# SNP Data
Source: SNPKIT variant calling
NFS Path: /nfs/turbo/umms-esnitkin/Project_Penn_KPC/Sequence_data/2024-05-07_SNPKIT_ST258/output_files/2024_05_07_16_13_40_core_results/matrices/SNP*
```{r}
## Load
SNP_code <- fread(file="../../../Sequence_data/variant_calling/2025-02-17_snpkit_Penn_KPC_ST258/output_files/2025_02_23_21_11_49_core_results/data_matrix/matrices/SNP_matrix_code.tsv")
SNP_allele <- fread(file="../../../Sequence_data/variant_calling/2025-02-17_snpkit_Penn_KPC_ST258/output_files/2025_02_23_21_11_49_core_results/data_matrix/matrices/SNP_matrix_allele_new.tsv")

## SNP parse
SNP_parsed <- clean_snpkit_output_SNP(SNP_code,SNP_allele,tr) 

# Sift
SIFT_scores_SNP <- fread("./data/SIFT/results/all_snp_SIFTannotations.xls") %>% `colnames<-`(c("chrom","pos","ref","var","transcript_id","locus_tag","gene_name","region","variant_type_sift","aa_ref","aa_var","aa_pos","sift_score","sift_median","num_seqs","dbSNP","sift_prediction")) %>% mutate(pos = as.numeric(pos)) %>% select(-chrom,-transcript_id,-gene_name)

# Merge 
rownames(SNP_parsed$annots) <- NULL
SNP_parsed$annots$pos <- as.numeric(SNP_parsed$annots$pos )  
SNP_parsed$annots <- left_join(SNP_parsed$annots,SIFT_scores_SNP)
saveRDS(SNP_parsed$bin,"./data/variant_matrices/SNP/SNP_mat.RDS")
saveRDS(SNP_parsed$annots,"./data/variant_matrices/SNP/SNP_annots.RDS")
```

# INDEL Data
Source: SNPKIT variant calling
NFS Path: /nfs/turbo/umms-esnitkin/Project_Penn_KPC/Sequence_data/2024-05-07_SNPKIT_ST258/output_files/2024_05_07_16_13_40_core_results/matrices/SNP*

```{r}
## Load
Indel_code <- fread(file="../../../Sequence_data/variant_calling/2025-02-17_snpkit_Penn_KPC_ST258/output_files/2025_02_23_21_11_49_core_results/data_matrix/matrices/Indel_matrix_code.tsv")
Indel_allele <- fread(file="../../../Sequence_data/variant_calling/2025-02-17_snpkit_Penn_KPC_ST258/output_files/2025_02_23_21_11_49_core_results/data_matrix/matrices/Indel_matrix_allele.tsv")

## Indel parse
INDEL_parsed <- clean_snpkit_output_INDEL(Indel_code,Indel_allele,tr) 

SIFT_scores_INDEL <- fread("./data/SIFT/results/all_indel_SIFTannotations.xls") %>% `colnames<-`(c("chrom","pos","ref","var","transcript_id","locus_tag","gene_name","region","variant_type_sift","aa_ref","aa_var","aa_pos","sift_score","sift_median","num_seqs","dbSNP","sift_prediction")) %>% mutate(pos = as.numeric(pos)) %>% select(-chrom,-transcript_id,-gene_name)

# Merge annotations
rownames(INDEL_parsed$annots) <- NULL
INDEL_parsed$annots$pos <- as.numeric(INDEL_parsed$annots$pos)  

INDEL_parsed$annots <- left_join(INDEL_parsed$annots,SIFT_scores_INDEL)

saveRDS(INDEL_parsed$bin,"./data/variant_matrices/INDEL/INDEL_mat.RDS")
saveRDS(INDEL_parsed$annots,"./data/variant_matrices/INDEL/INDEL_annots.RDS")
```

# Clean IS Data
Source: PanISA and ISFinder results generated by ISScreener pipeline
Path: 
```{r Clean IS Data}
# Curate reference genome
kpnih1 <- readGenBank(GBAccession("NZ_CP008827.1"),partial=TRUE) 
saveRDS(kpnih1,"./data/references/KPNIH1/KPNIH1_gbk_obj.RDS")

# IS finder curation
is_finder <- read.delim(file="./data/panISa/ISFinder_master.txt")  
is_data <-  parse_is(is_finder,gb = kpnih1 ,accession=F)

# Matrix
is_matrix <- is_data$mat  %>% .[,match(tr$tip.label,colnames(.))] %>% setNames(.,colnames(.))  %>% as.data.frame()

# Annotation dataset
is_annots <- is_data[names(is_data) %in% "mat" ==F] %>% bind_rows() %>% mutate(raw_rownames = rownames(is_data$mat))

# Create ST258 Matrix & Annotation Set
is_parsed <- remove_rows_with_no_variants_in_subset_unparsed(is_matrix,is_annots)
saveRDS(is_parsed$bin, file = paste0("./data/variant_matrices/IS/ISFinder_mat.RDS"))
saveRDS(is_parsed$annots, file = paste0("./data/variant_matrices/IS/ISFinder_annots.RDS"))
```

# All genomic data
```{r}
all_variants <- left_join(SNP_parsed$bin %>% mutate(isolate_no = rownames(.)),INDEL_parsed$bin%>% mutate(isolate_no = rownames(.))) %>% left_join(.,is_parsed$bin %>% mutate(isolate_no = rownames(.))) %>% `rownames<-`(.$isolate_no) %>% select(-isolate_no)
all_variants_dist_mat <- ape::dist.gene(x=all_variants,method = "pairwise",pairwise.deletion = T) %>%  as.matrix 
all_annots <- bind_rows(SNP_parsed$annots,INDEL_parsed$annots) %>% bind_rows(is_parsed$annots)

saveRDS(all_variants,"./data/variant_matrices/all_variants/all_variant_mat.RDS")
saveRDS(all_variants_dist_mat,"./data/nearest_neighbor_comparisons/all_variant_dist_mat.RDS")
saveRDS(all_annots,"./data/variant_matrices/all_variants/all_variant_annots.RDS")
```