---
title: "results_section_5"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r environment, echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse","gridExtra",'kableExtra',"tableone","ape","ggtree")

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")
source("./lib/common_functions.R")
```

```{r} 
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
# GWAS table
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS")
gwas_mat <- readRDS("./data/GWAS/hits/gwas_mat.RDS")
# Genome features
KPNIH1_features <- readRDS("./data/references/KPNIH1/KPNIH1_features.RDS")
# Matrices
## Kleborate matrix
kleborate <- readRDS("./data/GWAS/matrices/kleborate_mat.RDS") %>% as.data.frame 
## KPC Plasmids 
KPC_info_clusters <- readRDS("./data/KPC_plasmid/KPC_containing_clusters_mat.RDS") %>% as.data.frame 
KPC_containing_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_KPC")) %>% {ifelse(.=="KPC Plasmid",1,0)} %>% as.data.frame%>% select_if(colSums(.)>0)
KPC_clusters <- colnames(KPC_containing_clusters)
Non_KPC_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_non_KPC")) %>% {ifelse(.=="Non-KPC plasmid",1,0)} %>% as.data.frame %>% select_if(colSums(.)>0)
df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.))) %>% left_join(Non_KPC_clusters %>% mutate(isolate_no = rownames(.)))
## Core genome burden models
models <- c("model1","model1.2","model2","model2.2","model3","model3.2")
for(x in models){
  grouped_model <- paste0(x,"_burden.RDS")
  assign(gsub(".RDS","",grouped_model),readRDS(paste0("./data/GWAS/matrices/",grouped_model)) %>% as.data.frame)
}
## BLBLI clustering
pclustering <- readRDS('./data/asr_clustering/blbli_asr_clustering_df.RDS')
blbli_asr <- readRDS('./data/asr_clustering/blbli_asr.RDS')
## Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
## Depth
depth <- readRDS("./data/KPC_coverage/normalized/KPNIH1_KPC_plasmid_depth_by_feature_median_chromosome_normalized.RDS")
# Join
df <- left_join(df,kleborate %>% mutate(isolate_no = rownames(.)))
df <- left_join(df,gwas_mat %>% mutate(isolate_no = rownames(.))) 
df <- left_join(df,pclustering %>% select(isolate_no,blbli_asr_cluster_renamed))
df$blbli_asr_cluster_renamed_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible" )
df <- left_join(df,canonical %>% mutate(isolate_no = rownames(.))) 
df <- left_join(df,depth %>% mutate(isolate_no = rownames(.)))
```

# Log-2 KPC association

## OmpK36 GD insertion story
```{r} 
convert_tableone_into_df(df,c("OmpK36_L3_mutations","OmpK36GD","OmpK36TD"),strata="clade_I",outcome_names = c("Clade I","Clade II"),factorVars = c("OmpK36_L3_mutations","OmpK36GD","OmpK36TD"))
 
# Simple ompK36 status
df$OmpK36_mutations_simple <- ifelse(df$OmpK36_putative_function_altering==1 | df$OmpK36_truncation_kleborate==1 ,"Putative function-altering variant (PFAV)",
                                     ifelse(df$OmpK36_L3_mutations==1, "Loop 3 insertion","No loop 3 insertion or PFAV"))

df$OmpK36_mutations_simple <- factor(df$OmpK36_mutations_simple, levels=rev(c("No loop 3 insertion or PFAV","Loop 3 insertion","Putative function-altering variant (PFAV)")))
# Easy variable
df$KPC_log_2 <- log2(df$KPNIH1_RS28775_chr_median_norm)
```

```{r}
df_w_KPC <- df %>% subset(KPNIH1_RS28775_chr_median_norm >0)
df_w_KPC$OmpK36_mutations_simple <- relevel(df_w_KPC$OmpK36_mutations_simple,ref = "No loop 3 insertion or PFAV") 
paste0("KPC w/ MVB or IR")
lm(MVB_log_2 ~ KPC_log_2 ,data=df_w_KPC) %>% summary 
lm(MVB_log_2 ~ KPC_log_2 + OmpK36_mutations_simple,data=df_w_KPC) %>% summary  
lm(MVB_log_2 ~ KPC_log_2 * OmpK36_mutations_simple,data=df_w_KPC) %>% summary 

paste0("IR w/ KPC")
lm(IR_log_2 ~ KPC_log_2 ,data=df_w_KPC) %>% summary 
lm(IR_log_2 ~ KPC_log_2 + OmpK36_mutations_simple,data=df_w_KPC) %>% summary  
lm(IR_log_2 ~ KPC_log_2 * OmpK36_mutations_simple,data=df_w_KPC) %>% summary  
```
### Investigation across clades
```{r}
df_w_KPC_CI <- df_w_KPC %>% subset(clade_I == "Clade I")
df_w_KPC_CII <- df_w_KPC %>% subset(clade_I != "Clade I")
# Clade I
paste0("KPC w/ MVB or IR")
lm(MVB_log_2 ~ KPC_log_2 ,data=df_w_KPC_CI) %>% summary 
lm(MVB_log_2 ~ KPC_log_2 + OmpK36_mutations_simple,data=df_w_KPC_CI) %>% summary  
lm(MVB_log_2 ~ KPC_log_2 * OmpK36_mutations_simple,data=df_w_KPC_CI) %>% summary 


paste0("IR w/ KPC")
lm(IR_log_2 ~ KPC_log_2 ,data=df_w_KPC_CI) %>% summary 
lm(IR_log_2 ~ KPC_log_2 + OmpK36_mutations_simple,data=df_w_KPC_CI) %>% summary  
lm(IR_log_2 ~ KPC_log_2 * OmpK36_mutations_simple,data=df_w_KPC_CI) %>% summary  
```
