---
title: "Convergence-based genome-wide association study identified additional resistance-associated genotypes"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_mount/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/')
```

```{r environment, echo=F, message=FALSE, include=F, results=F}
#Packages
packages <- c("tidyverse","gridExtra",'kableExtra',"tableone","ape","ggtree")

#Load Packages
invisible(lapply(packages,library,character.only=T,quietly=T))

#Scripts
source("./lib/consistent_themes.R")
source("./scripts/GWAS/analysis/GWAS_analysis_scripts.R")
source("./lib/common_functions.R")
```

```{r} 
# Standard data
df <- readRDS("./data/dataset/df.RDS")
tr <- read.tree("./data/tree/tree.treefile")
# GWAS table
gwas_table <- readRDS("./data/GWAS/hits/gwas_hits_table.RDS")
gwas_mat <- readRDS("./data/GWAS/hits/gwas_mat.RDS")
# Genome features
KPNIH1_features <- readRDS("./data/references/KPNIH1/KPNIH1_features.RDS")
# Matrices
## Kleborate matrix
kleborate <- readRDS("./data/GWAS/matrices/kleborate_mat.RDS") %>% as.data.frame 
## KPC Plasmids 
KPC_info_clusters <- readRDS("./data/KPC_plasmid/KPC_containing_clusters_mat.RDS") %>% as.data.frame 
KPC_containing_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_KPC")) %>% {ifelse(.=="KPC Plasmid",1,0)} %>% as.data.frame%>% select_if(colSums(.)>0)
KPC_clusters <- colnames(KPC_containing_clusters)
Non_KPC_clusters <- KPC_info_clusters %>% `colnames<-`(paste0(colnames(.),"_non_KPC")) %>% {ifelse(.=="Non-KPC plasmid",1,0)} %>% as.data.frame %>% select_if(colSums(.)>0)
df <- left_join(df,KPC_containing_clusters %>% mutate(isolate_no = rownames(.))) %>% left_join(Non_KPC_clusters %>% mutate(isolate_no = rownames(.)))
## Core genome burden models
models <- c("model1","model1.2","model2","model2.2","model3","model3.2")
for(x in models){
  grouped_model <- paste0(x,"_burden.RDS")
  assign(gsub(".RDS","",grouped_model),readRDS(paste0("./data/GWAS/matrices/",grouped_model)) %>% as.data.frame)
}
## BLBLI clustering
pclustering <- readRDS('./data/asr_clustering/blbli_asr_clustering_df.RDS')
blbli_asr <- readRDS('./data/asr_clustering/blbli_asr.RDS')
## Canonical
canonical <- readRDS("./data/carbapenem_resistance_panel/carbapenem_resistance_panel_matrix.RDS")
## Depth
depth <- readRDS("./data/KPC_coverage/normalized/KPNIH1_KPC_plasmid_depth_by_feature_median_chromosome_normalized.RDS")
# Join
df <- left_join(df,kleborate %>% mutate(isolate_no = rownames(.)))
df <- left_join(df,gwas_mat %>% mutate(isolate_no = rownames(.))) 
df <- left_join(df,pclustering %>% select(isolate_no,blbli_asr_cluster_renamed))
df$blbli_asr_cluster_renamed_simple <- recode(df$blbli_asr_cluster_renamed,"No Feature"="Susceptible" )
   
df <- left_join_matrix(model1_burden,df,"_1")
df <- left_join_matrix(model1.2_burden,df,"_1.2")
df <- left_join_matrix(model2_burden,df,"_2")
df <- left_join_matrix(model2.2_burden,df,"_2.2")
df <- left_join_matrix(model3_burden,df,"_3")
df <- left_join_matrix(model3.2_burden,df,"_3.2")
df <- left_join(df,canonical %>% mutate(isolate_no = rownames(.))) 
df <- left_join(df,depth %>% mutate(isolate_no = rownames(.)))

# Easy variable for blaKPC copy numebr
df$KPC_log_2 <- log2(df$KPNIH1_RS28775_chr_median_norm)
```

# Overview
```{r}
paste0("Number of gwas hits: ", nrow(gwas_table))

paste0("Resistance category:")
table(gwas_table$resistance_category)

paste0("How many passed QC:")
table(gwas_table$nn_qc)
```

# Four modulators
```{r}
continuous <- gwas_table %>% subset(resistance_category == "Modulator")

paste0("Number of continuous hits: ")
nrow(continuous)

paste0("Number of continuous hits passing qc: ")
continuous_pass <- subset(continuous, nn_qc=="Yes")
nrow(continuous_pass)

paste0("Genotypes passing:") 
library(rstatix)
enrichment <- function(group,df){
  group_df <- df %>% subset(get(group)==1)
  wo_df <- df %>% subset(get(group)==0)
  # MVB
  ## Median IQR
  MVB_median_with <-  median(group_df[,'MVB_log_2'])
  MVB_IQR_with <-  IQR(group_df[,'MVB_log_2'])
  MVB_median_IQR_with <-  paste0(MVB_median_with," (",MVB_IQR_with,")")  
  MVB_median_wo <-  median(wo_df[,'MVB_log_2'])
  MVB_IQR_wo <-  IQR(wo_df[,'MVB_log_2'])
  MVB_median_IQR_wo <-  paste0(MVB_median_wo," (",MVB_IQR_wo,")") 
  ## T test
  MVB_t_test <- wilcox.test(df[,group], df[,"MVB_log_2"], conf.int=TRUE)    
  MVB_t_test_p <- MVB_t_test$p.value
  MVB_t_test_estimate <- wilcox_effsize(data = df,formula = as.formula(paste0("MVB_log_2","~",group)),paired = F) %>% .$effsize
  
  # IR
  ## Median IQR
  IR_median_with <-  median(group_df[,'IR_log_2'])
  IR_IQR_with <-  IQR(group_df[,'IR_log_2'])
  IR_median_IQR_with <-  paste0(IR_median_with," (",IR_IQR_with,")")  
  IR_median_wo <-  median(wo_df[,'IR_log_2'])
  IR_IQR_wo <-  IQR(wo_df[,'IR_log_2'])
  IR_median_IQR_wo <-  paste0(IR_median_wo," (",IR_IQR_wo,")")  
  ## T test
  IR_t_test <- wilcox.test(df[,group], df[,"IR_log_2"], conf.int=TRUE) 
  IR_t_test_p <- IR_t_test$p.value
  IR_t_test_estimate <- wilcox_effsize(data = df,formula = as.formula(paste0("IR_log_2","~",group)),paired = F) %>% .$effsize
  return(data.frame(genotype=group,MVB_median_IQR_with,MVB_median_IQR_wo,MVB_t_test_estimate,MVB_t_test_p,IR_median_IQR_with,IR_median_IQR_wo,IR_t_test_estimate,IR_t_test_p))
}

lapply(continuous_pass$genotype,FUN=enrichment,df) %>% do.call(rbind,.) %>% favorite_kable
  

continuous_pass %>% select(locus_tag,product,MVB_num_log_2_diff_increase_prop,MVB_num_log_2_diff_decrease_prop,MVB_num_log_2_diff_median,MVB_num_log_2_diff_range,IR_num_log_2_diff_increase_prop,IR_num_log_2_diff_decrease_prop,IR_num_log_2_diff_median,IR_num_log_2_diff_range,genotype,sig_tests,sig_phenotypes)  %>% favorite_kable
```

# Synchronous hits
```{r}
synchronous <- gwas_table %>% subset(resistance_category != "Modulator")

paste0("Number of synchronous hits")
nrow(synchronous)

paste0("Synchronous hits that passed QC")
synchronous_pass <- subset(synchronous, nn_qc=="Yes")

synchronous_pass %>% select(locus_tag,gene,product,blbli_res_freq,blbli_sus_freq,`Singleton_prop`,blbli_dich_num_binary_diff_gain_prop,blbli_dich_num_binary_diff_loss_prop,best_model,genotype)  %>% favorite_kable

novel_synchronous <- synchronous_pass %>% subset(!locus_tag %in% c("KPNIH1_RS18665","KPNIH1_RS07805")) %>% .$genotype
```

# AA552 plasmid
```{r} 
plasmid_table <- table(df$AA552_KPC,df$blbli_dich_num) 
plasmid_table
prop.table(plasmid_table,margin=2)
chisq.test( plasmid_table) 
```

```{r}
df_CI <- subset(df,clade_I == 'Clade I')
convert_tableone_into_df(df_CI,vars=c("AA018","AA552"),factorVars = c("AA018","AA552"))

enrichment("AA552",df_CI)

wilcox.test(df_CI[,"AA552"],df_CI[,"KPC_log_2"],alternative='two.sided')
```

```{r}
table(df_CI$AA018_KPC,df_CI$AA552_non_KPC)
convert_tableone_into_df(df_CI,"AA552_non_KPC",strata="AA018_KPC",factorVars = "AA552_non_KPC",outcome_names = c("Absent","Present"))
convert_tableone_into_df(df_CI,"AA018_non_KPC",strata="AA552_KPC",factorVars = "AA018_non_KPC",outcome_names = c("Absent","Present"))
```